keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -430,32 +438,47 @@ Object:
   properties:
     invalid:
       type: string
-      default: default-name
+      default: "default-name"
+      # This field would only be present for CRDs, for built-in,
+      # the empty value would be accepted and treated as missing.
+      minLength: 1
     defaulted:
-      type: boolean
+      type: int
       default: false
 ```
 
 Which, for both CRD types and built-in types, would generate the following behavior:
 ```python
 >>> default('{}')
 {"entry": {"name": "default-name", "number": 0}}
->>> default('{entry: {name: other-name}}')
-{"entry": {"name": "other-name", "number": 0}}
+>>> default('{name: other-name}')
+{"name": "other-name", "number": 0}
+# For built-in, additionally, we would have:
+>>> default('{name: ""}')
+{"name": "default-name", "number": 0}
 ```
 
-Note that the generated code for built-in will look like this:
+Note that a simplified, and non-cached version of the generated code for
+built-ins would look like this:
 ```golang
 func GenerateDefault(obj *Object) {
      if obj.Name == "" {
-          obj.Name = "default-name"
+          if err := json.Unmarshal([]byte(`"default-name"`), &obj.Name); err != nil {
+               panic(err)
+          }
      }
      if obj.Number == 0 {
-          obj.Number = 0
+          if err := json.Unmarshal([]byte(`0`), &obj.Number); err != nil {
+               panic(err)
+          }
      }
 }
 ```
 
+It's still possible to specify an invalid zero-value in built-in types
+embedded into a CRD, but client-go will drop these values, thanks to
+`omitempty`.
```

Comment:
What do you mean? Why does client-go matter for API semantics?

Reply:
I agree, it won't solve the api semantics, but it will do the right thing as long as people use client-go. That's already true for many things (you can specify `protocol: ""` and it should probably be an error, but client-go fixes it for you)